### ISSUE TITLE
[Question] How is the similarity matrix calculated using PPR?

### ISSUE THREAD

**Issue by aj280192** [xgfs/verse#12]
Hi Authors,

I am trying to understand how the similarity matrix using PPR is calculated in the implementation. Which part of the source code is implementing the similarity matrix calculation using PPR(my C++ knowledge is not that good)? Most of the resources for PPR calculation return 1 × V  but as per your paper, the similarity matrix should be V × V → R. So I am trying to understand how it was calculated to be V × V using PPR. I am trying to implement the algorithm using Pytorch and DGL. Any help is much appreciated.

**Comment by xgfs** [comment]
Hey,

Version implemented here is 1 x V version of the algorithm, since it works faster and requires less tuning in practice. We do have a full V x V version in C++, but it's not as neat and works slowly. If you drop me an email I can send the code, but I'd rather not open-source it, as I can not provide timely support.

**Comment by aj280192** [comment]
Thanks for the immediate reply!. So this changes the working of single-layer neural network right? from V neurons on input-layer d on hidden-layer and finally V output neurons to 1 -> d -> 1 right?

**Comment by xgfs** [comment]
I mean, there is only one weight matrix we are optimizing (equivalent to one torch.nn.Embedding), so you would fetch a source node, a target node (end point of a random walk), and k random negatives. Then you'd multiply 1 x v vector by [v x 1+k] to get 1+k scores, and these can be fed to binary crossentropy loss. I would assume this would be much slower than this version, though. Happy to be proven wrong :)

**Comment by aj280192** [comment]
I have dropped an email to you!! anton@tsitsul.in.
